File: conformance/runtime/ts/carrier.ts
// Selects the bitruntime carrier a conformance driver runs on, as ../go's
// internal/carrier does: a local pair, or a real WebSocket between two
// bitruntime peers whose client or server side sends. It supplies only
// construction and cleanup; every delivery is bitruntime's own.
import { once } from 'node:events';
import { WebSocket, WebSocketServer } from 'ws';
import type { Endpoint } from '@bitspark/bitwire';
import { pair } from '@bitspark/bitruntime/core';
import { Peer } from '@bitspark/bitruntime/engine';
import { CODE_NORMAL, webSocketConnection, type WebSocketLike } from '@bitspark/bitruntime/transports';

/** BITRUNTIME_CARRIER is local or peer; BITRUNTIME_REVERSE=1 makes a peer's server side the sending origin. */
export function kind(): 'local' | 'peer' {
  const carrier = process.env.BITRUNTIME_CARRIER;
  if (carrier !== 'local' && carrier !== 'peer') throw new Error('Expected BITRUNTIME_CARRIER=local or peer');
  if (!['', '0', '1'].includes(process.env.BITRUNTIME_REVERSE ?? '')) throw new Error('Expected BITRUNTIME_REVERSE=0 or 1');
  return carrier;
}

// ws implements the browser event surface the adapter uses; its overloaded
// addEventListener declaration has a different TypeScript shape.
function asLike(socket: WebSocket): WebSocketLike {
  socket.binaryType = 'arraybuffer';
  return socket as unknown as WebSocketLike;
}

/**
 * Returns two connected endpoints: sending on the first delivers to the
 * receiver of the second, and the reverse. cleanup registers each release.
 */
export async function connected(cleanup: (release: () => void) => void): Promise<[Endpoint, Endpoint]> {
  if (kind() === 'local') {
    const [a, b] = pair();
    cleanup(() => {
      a.close(CODE_NORMAL, 'done');
      b.close(CODE_NORMAL, 'done');
    });
    return [a, b];
  }
  const listener = new WebSocketServer({ port: 0, host: '127.0.0.1' });
  cleanup(() => listener.close());
  await once(listener, 'listening', { signal: AbortSignal.timeout(5_000) });
  const address = listener.address();
  if (!address || typeof address === 'string') throw new Error('No WebSocket address');
  const accepted = once(listener, 'connection', { signal: AbortSignal.timeout(5_000) });
  // Both awaits can fail independently; keep a rejection handler while the
  // client's handshake is pending, then await this same promise.
  accepted.catch(() => {});
  listener.on('connection', (socket: WebSocket) => cleanup(() => socket.terminate()));
  const socket = new WebSocket(`ws://127.0.0.1:${address.port}`);
  cleanup(() => socket.terminate());
  await once(socket, 'open', { signal: AbortSignal.timeout(5_000) });
  const [remote] = (await accepted) as [WebSocket];
  const client = new Peer({ role: 'client' });
  const server = new Peer({ role: 'server' });
  cleanup(() => {
    client.close();
    server.close();
  });
  await Promise.all([client.attach(webSocketConnection(asLike(socket))), server.attach(webSocketConnection(asLike(remote)))]);
  return process.env.BITRUNTIME_REVERSE === '1' ? [server.wire(), client.wire()] : [client.wire(), server.wire()];
}

File: conformance/runtime/go/internal/carrier/carrier.go
// Package carrier selects the bitruntime carrier a conformance driver runs on,
// exactly as the Nightseam baseline selects its carriers: a local pair, or a
// WebSocket connection whose client or server side sends. It supplies only
// construction and cleanup; every delivery is bitruntime's own.
package carrier

import (
	"context"
	"net/http"
	"net/http/httptest"
	"os"
	"time"

	core "github.com/Bitspark/bitruntime/core/go"
	engine "github.com/Bitspark/bitruntime/engine/go"
	websocket "github.com/Bitspark/bitruntime/engine/websocket/go"
	transports "github.com/Bitspark/bitruntime/transports/go"
	wire "github.com/Bitspark/bitwire/wire/go"
)

// Kind is BITRUNTIME_CARRIER: "local" or "peer". A peer is a real WebSocket
// between two bitruntime engines; BITRUNTIME_REVERSE=1 makes the server side
// the sending origin.
func Kind() string {
	kind := os.Getenv("BITRUNTIME_CARRIER")
	if kind != "local" && kind != "peer" {
		panic("expected BITRUNTIME_CARRIER=local or peer")
	}
	if reverse := os.Getenv("BITRUNTIME_REVERSE"); reverse != "" && reverse != "0" && reverse != "1" {
		panic("expected BITRUNTIME_REVERSE=0 or 1")
	}
	return kind
}

func check(err error) {
	if err != nil {
		panic(err)
	}
}

// Pair returns two connected endpoints. Sending on the first delivers to the
// receiver of the second, and the reverse. cleanup registers each release.
func Pair(cleanup func(func())) (wire.Endpoint, wire.Endpoint) {
	if Kind() == "local" {
		a, b, err := core.NewPair(core.PairOptions{})
		check(err)
		cleanup(func() {
			_ = a.Close(transports.CodeNormal, "done")
			_ = b.Close(transports.CodeNormal, "done")
		})
		return a, b
	}
	connected := make(chan *engine.Peer, 1)
	handler, err := websocket.NewHandler(websocket.ServerOptions{
		Authenticate: func(r *http.Request) (context.Context, error) { return r.Context(), nil },
		CheckOrigin:  func(*http.Request) bool { return true },
		OnConnect:    func(peer *engine.Peer) { connected <- peer },
	})
	check(err)
	server := httptest.NewServer(handler)
	cleanup(server.Close)
	ctx, cancel := context.WithTimeout(context.Background(), time.Minute)
	cleanup(cancel)
	client, _, err := websocket.Dial(ctx, server.URL, websocket.DialOptions{ConnectTimeout: 5 * time.Second})
	check(err)
	cleanup(func() { _ = client.Close() })
	var remote *engine.Peer
	select {
	case remote = <-connected:
	case <-time.After(5 * time.Second):
		panic("the server did not publish its accepted peer")
	case <-ctx.Done():
		panic(ctx.Err())
	}
	cleanup(func() { _ = remote.Close() })
	if os.Getenv("BITRUNTIME_REVERSE") == "1" {
		return remote.Wire(), client.Wire()
	}
	return client.Wire(), remote.Wire()
}
